kiva/blend2d.py

```diff
@@ -0,0 +1,780 @@
+            self.gc.fill_path(self.path)
+        if mode in (constants.STROKE, constants.FILL_STROKE):
+            self.gc.stroke_path(self.path)
+        self.begin_path()
+
+    def get_empty_path(self):
+        """ Return a path object that can be built up and then reused.
+        """
+        return CompiledPath()
+
+    def draw_path_at_points(self, points, path, mode=constants.FILL_STROKE):
+        """ Draw a path object at many different points.
+        """
+        raise NotImplementedError()
+
+    def draw_marker_at_points(self, points_array, size,
+                              marker=constants.SQUARE_MARKER):
+        """ Draw a marker at a collection of points
+        """
+        raise NotImplementedError()
+
+    def save(self, filename, file_format=None, pil_options=None):
+        """ Save the contents of the context to a file
+        """
+        if file_format is None:
+            file_format = ""
+        if pil_options is None:
+            pil_options = {}
+
+        img = self.to_image()
+
+        ext = (
+            os.path.splitext(filename)[1][1:] if isinstance(filename, str)
+            else ""
+        )
+
+        # Check the output format to see if it can handle an alpha channel.
+        no_alpha_formats = ("jpg", "bmp", "eps", "jpeg")
+        if ext in no_alpha_formats or file_format.lower() in no_alpha_formats:
+            img = img.convert("RGB")
+
+        # Check the output format to see if it can handle DPI
+        dpi_formats = ("jpg", "png", "tiff", "jpeg")
+        if ext in dpi_formats or file_format.lower() in dpi_formats:
+            # Assume 72dpi is 1x
+            dpi = int(72 * self.base_scale)
+            pil_options["dpi"] = (dpi, dpi)
+
+        img.save(filename, format=file_format, **pil_options)
+
+    def to_image(self):
+        """ Return the contents of the context as a PIL Image.
+
+        If the graphics context is in BGRA format, it will convert it to
+        RGBA for the image.
+
+        Returns
+        -------
+        img : Image
+            A PIL/Pillow Image object with the data in RGBA format.
+        """
+        try:
+            from PIL import Image
+        except ImportError:
+            raise ImportError("need Pillow to save images")
+
+        # Data is BGRA; Convert to RGBA
+        data = np.empty(self._buffer.shape, dtype=np.uint8)
+        data[..., 0] = self._buffer[..., 2]
+        data[..., 1] = self._buffer[..., 1]
+        data[..., 2] = self._buffer[..., 0]
+        data[..., 3] = self._buffer[..., 3]
+
+        return Image.fromarray(data, "RGBA")
+
+
+class CompiledPath(object):
+    def __init__(self):
+        self.path = blend2d.Path()
+
+    def copy(self):
+        raise NotImplementedError()
+
+    def begin_path(self):
+        self.path.clear()
+
+    def move_to(self, x, y):
+        self.path.move_to(x, y)
+
+    def arc(self, x, y, r, start_angle, end_angle, clockwise=False):
```

Comment:
nit : looks like agg uses `cw` instead of `clockwise` for this interface. Again, not sure what we want to do about it - if we want to switch this to `cw` or if we want to force everyone to use `clockwise` (i'd prefer the latter but that might be a lengthy deprecation/move)

Reply:
This backend started as a copy of `celiagg` (which was a copy of `qpainter`) so we'll need to unify those. This should be added to #648